File: Runtime/IntervalMeanTimestepMap.vb
'************************************************************************************
' ST-Sim: A .NET class library for developing state-and-transition simulation models.
'
' Copyright © 2007-2017 Apex Resource Management Solution Ltd. (ApexRMS). All rights reserved.
'
'************************************************************************************

''' <summary>
''' IntervalMeanTimestepMap
''' </summary>
''' <remarks></remarks>
Class IntervalMeanTimestepMap

    Private m_Map As New Dictionary(Of Integer, Integer)
    Private m_MinimumTimestep As Integer
    Private m_MaximumTimestep As Integer
    Private m_TimestepZero As Integer
    Private m_Frequency As Integer

    Public Sub New(
        ByVal minimumTimestep As Integer,
        ByVal maximumTimestep As Integer,
        ByVal timestepZero As Integer,
        ByVal frequency As Integer)

        Me.m_MinimumTimestep = minimumTimestep
        Me.m_MaximumTimestep = maximumTimestep
        Me.m_TimestepZero = timestepZero
        Me.m_Frequency = frequency

        Me.FillMap()

    End Sub

    Public Function GetValue(ByVal value As Integer) As Integer
        Return Me.m_Map(value)
    End Function

    Private Sub FillMap()

        'Handle special cases that are relatively common and/or very simple
        If (Me.m_MinimumTimestep = Me.m_MaximumTimestep) Then

            Me.m_Map.Add(Me.m_MinimumTimestep, Me.m_MaximumTimestep)

        ElseIf (Me.m_Frequency = 1) Then

            For CurrentTimestep As Integer = Me.m_MinimumTimestep To Me.m_MaximumTimestep
                Me.m_Map.Add(CurrentTimestep, CurrentTimestep)
            Next

        Else

            Dim CurrentTimestep As Integer = Me.m_MinimumTimestep

            While (CurrentTimestep <= Me.m_MaximumTimestep)

                Dim AggregatorTimestep As Integer = Me.GetNextAggregatorTimestep(CurrentTimestep)
                Debug.Assert(AggregatorTimestep <= Me.m_MaximumTimestep)

                While (CurrentTimestep <= AggregatorTimestep)

                    Me.m_Map.Add(CurrentTimestep, AggregatorTimestep)
                    CurrentTimestep += 1

                End While

            End While

        End If

#If DEBUG Then

        Debug.Assert(Me.m_Map.Count = (Me.m_MaximumTimestep - Me.m_MinimumTimestep + 1))

        For t As Integer = Me.m_MinimumTimestep To Me.m_MaximumTimestep
            Debug.Assert(Me.m_Map.ContainsKey(t))
        Next

#End If

    End Sub

    ''' <summary>
    ''' Gets the next 'Aggregator' timestep
    ''' </summary>
    ''' <param name="currentAggregatorTimestep"></param>
    ''' <returns></returns>
    ''' <remarks>This function exists to support the 'calculate as interval mean values' feature for summary transition output.</remarks>
    Private Function GetNextAggregatorTimestep(ByVal currentAggregatorTimestep As Integer) As Integer

        Dim NextAggregatorTimestep As Integer = currentAggregatorTimestep

        While (NextAggregatorTimestep < Me.m_MaximumTimestep)

            If (((NextAggregatorTimestep - Me.m_TimestepZero) Mod Me.m_Frequency) = 0) Then
                Return NextAggregatorTimestep
            End If

            NextAggregatorTimestep += 1

        End While

        Debug.Assert(NextAggregatorTimestep = Me.m_MaximumTimestep)
        Return NextAggregatorTimestep

    End Function

End Class

File: src/Runtime/IntervalMeanTimestepMap.cs
// stsim: A SyncroSim Package for developing state-and-transition simulation models using ST-Sim.
// Copyright © 2007-2021 Apex Resource Management Solutions Ltd. (ApexRMS). All rights reserved.

using System.Diagnostics;
using System.Collections.Generic;

namespace SyncroSim.STSim
{
    internal class IntervalMeanTimestepMap
    {
        private Dictionary<int, int> m_Map = new Dictionary<int, int>();
        private int m_MinimumTimestep;
        private int m_MaximumTimestep;
        private int m_TimestepZero;
        private int m_Frequency;

        public IntervalMeanTimestepMap(int minimumTimestep, int maximumTimestep, int timestepZero, int frequency)
        {
            this.m_MinimumTimestep = minimumTimestep;
            this.m_MaximumTimestep = maximumTimestep;
            this.m_TimestepZero = timestepZero;
            this.m_Frequency = frequency;

            this.FillMap();
        }

        public int GetValue(int value)
        {
            return this.m_Map[value];
        }

        private void FillMap()
        {
            //Handle special cases that are relatively common and/or very simple
            if (this.m_MinimumTimestep == this.m_MaximumTimestep)
            {
                this.m_Map.Add(this.m_MinimumTimestep, this.m_MaximumTimestep);
            }
            else if (this.m_Frequency == 1)
            {
                for (int CurrentTimestep = this.m_MinimumTimestep; CurrentTimestep <= this.m_MaximumTimestep; CurrentTimestep++)
                {
                    this.m_Map.Add(CurrentTimestep, CurrentTimestep);
                }
            }
            else
            {
                int CurrentTimestep = this.m_MinimumTimestep;

                while (CurrentTimestep <= this.m_MaximumTimestep)
                {
                    int AggregatorTimestep = this.GetNextAggregatorTimestep(CurrentTimestep);
                    Debug.Assert(AggregatorTimestep <= this.m_MaximumTimestep);

                    while (CurrentTimestep <= AggregatorTimestep)
                    {
                        this.m_Map.Add(CurrentTimestep, AggregatorTimestep);
                        CurrentTimestep += 1;
                    }
                }
            }

#if DEBUG

            Debug.Assert(this.m_Map.Count == (this.m_MaximumTimestep - this.m_MinimumTimestep + 1));

            for (int t = this.m_MinimumTimestep; t <= this.m_MaximumTimestep; t++)
            {
                Debug.Assert(this.m_Map.ContainsKey(t));
            }

#endif
        }

        /// <summary>
        /// Gets the next 'Aggregator' timestep
        /// </summary>
        /// <param name="currentAggregatorTimestep"></param>
        /// <returns></returns>
        /// <remarks>This function exists to support the 'calculate as interval mean values' feature for summary transition output.</remarks>
        private int GetNextAggregatorTimestep(int currentAggregatorTimestep)
        {
            int NextAggregatorTimestep = currentAggregatorTimestep;

            while (NextAggregatorTimestep < this.m_MaximumTimestep)
            {
                if (((NextAggregatorTimestep - this.m_TimestepZero) % this.m_Frequency) == 0)
                {
                    return NextAggregatorTimestep;
                }

                NextAggregatorTimestep += 1;
            }

            Debug.Assert(NextAggregatorTimestep == this.m_MaximumTimestep);
            return NextAggregatorTimestep;
        }
    }
}
